Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/Async/AsyncQPSClientImpl.swift

```diff
@@ -0,0 +1,293 @@
+    self.eventLoopGroup = eventLoopGroup
+    
+    // Parse possible invalid targets before code with side effects.
+    let serverTargets = try config.parsedServerTargets()
+    precondition(serverTargets.count > 0)
+
+    // Start recording stats.
+    self.statsPeriodStart = grpcTimeNow()
+    self.cpuStatsPeriodStart = getResourceUsage()
+
+    let requestMessage = try AsyncQPSClientImpl.makeClientRequest(payloadConfig: config.payloadConfig)
+
+    // Start the requested number of channels.
+    self.channelRepeaters = (0 ..< Int(config.clientChannels)).map { channelNumber in
+      ChannelRepeater(
+        target: serverTargets[channelNumber % serverTargets.count],
+        requestMessage: requestMessage,
+        config: config,
+        eventLoop: eventLoopGroup.next()
+      )
+    }
+  }
+  
+  /// Start the execution of the client.
+  func startClient() {
+    Task {
+      try await withThrowingTaskGroup(of: Void.self) { group in
+        for repeater in self.channelRepeaters {
+          group.addTask {
+            try await repeater.start()
+          }
+        }
+        try await group.waitForAll()
+      }
+    }
+  }
+
+  /// Send current status back to the driver process.
+  /// - parameters:
+  ///     - reset: Should the stats reset after being sent.
+  ///     - context: Calling context to allow results to be sent back to the driver.
+  func sendStatus(reset: Bool, responseStream: GRPCAsyncResponseStreamWriter<Grpc_Testing_ClientStatus>) async throws {
+    let currentTime = grpcTimeNow()
+    let currentResourceUsage = getResourceUsage()
+    var result = Grpc_Testing_ClientStatus()
+    result.stats.timeElapsed = (currentTime - self.statsPeriodStart).asSeconds()
+    result.stats.timeSystem = currentResourceUsage.systemTime - self.cpuStatsPeriodStart
+      .systemTime
+    result.stats.timeUser = currentResourceUsage.userTime - self.cpuStatsPeriodStart.userTime
+    result.stats.cqPollCount = 0
+
+    // Collect stats from each of the channels.
+    var latencyHistogram = Histogram()
+    var statusCounts = StatusCounts()
+    for channelRepeater in self.channelRepeaters {
+      let stats = channelRepeater.getStats(reset: reset)
+      try! latencyHistogram.merge(source: stats.latencies)
+      statusCounts.merge(source: stats.statuses)
+    }
+    result.stats.latencies = Grpc_Testing_HistogramData(from: latencyHistogram)
+    result.stats.requestResults = statusCounts.toRequestResultCounts()
+    self.logger.info("Sending client status")
+    try await responseStream.send(result)
+
+    if reset {
+      self.statsPeriodStart = currentTime
+      self.cpuStatsPeriodStart = currentResourceUsage
+    }
+  }
+
+  /// Shutdown the service.
+  /// - parameters:
+  ///     - callbackLoop: Which eventloop should be called back on completion.
+  /// - returns: A future on the `callbackLoop` which will succeed on completion of shutdown.
+  func shutdown() async throws {
+    for repeater in self.channelRepeaters {
```

Comment:
We can use a task group here to shutdown the repeaters concurrently. At the moment we're waiting for each to stop before moving onto the next.